### ISSUE TITLE
[bug] test_autograd sometimes runs CUDA tests on CPU

### ISSUE THREAD

**Issue by kshitij12345** [pytorch/pytorch#53361]
Based on the offline discussion with @antocuni @nikitaved , we realized that few CUDA Autograd tests don't really run on CUDA.

Example: 
https://github.com/pytorch/pytorch/blob/854cc53594b4bb1871caad990e966e0de47abff8/torch/testing/_internal/common_methods_invocations.py#L3386-L3387

Here `dtype` and `device` arguments to the lambda are unused.

On digging for `random_square_matrix_of_rank` we see that,
https://github.com/pytorch/pytorch/blob/854cc53594b4bb1871caad990e966e0de47abff8/torch/testing/_internal/common_utils.py#L1590-L1599

`device` defaults to `cpu`.

--------------------------------------------

From cursory glance many entries have this issue
https://github.com/pytorch/pytorch/blob/854cc53594b4bb1871caad990e966e0de47abff8/torch/testing/_internal/common_methods_invocations.py#L3380-L3428

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @jianyuh @mruberry @heitorschueroff @walterddr @IvanYashchuk @VitalyFedyunin

**Comment by nikitaved** [comment]
To add on this, while `det` in forward is using a LU decomposition, the `det_backward` is using `svd_backward`, which is only double backward stable for inputs with distinct singular values. We will need to reimplement `det_backward` which uses `lu_backward` instead. `lu_backward` needs to be implemented in ATen, as the current backward for LU is done via `autograd.Function`. I will submit an issue about it.

**Comment by IvanYashchuk** [comment]
Yes, for this example `dtype` and `device` arguments to the lambda are unused and `random_square_matrix_of_rank` defaults to `'cpu'` and `double`. But then in `test_autograd.py` the input to gradchecks is created with `create_input` from `common_methods_invocations.py`.
`create_input` first calls the provided lambda
https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L3636-L3637
Let's say `dtype=cdouble`, `device='cuda'`, then for this example call to lambda gives a Tensor with `double` dtype and on `cpu`, but `map_arg` is called recursively so then it hits the other condition:
https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L3624-L3635
where the Tensor is cast to the specified `dtype` and transferred to the specified `device` and finally gradcheck receives an array with requested dtype and device.

**Comment by mruberry** [comment]
As we move forward let's be mindful of two design principles:

- use tensor creation helpers designed for the test suite like make_tensor() and random_symmetric_pd_matrix()
- ensure those helpers have required kwargs for device and dtype to make it hard to create a tensor on the wrong device or with the wrong dtype

We occasionally find other tests that aren't run on the correct device or dtype, too, simply because someone forgot to plug the given device and dtype into a constructor like torch.randn().

**Comment by antocuni** [comment]
@IvanYashchuk I think that your analysis is correct for `test_autograd.py` but **not** for `test_ops.py`, in particular `test_fn_grad` and `test_fn_gradgrad`.
In the following examples, I am referring to `linalg.det` in PR #53119, which shows the problem which triggered this issue. In particular, all the diffs I show are on top of commit 669a319eb0.

On my branch, `test_fn_gradgrad_linalg_det_cuda_float64` fails. After some trial&error, I reduced the problem to the following sample_input:
```python
def sample_inputs_linalg_det(op_info, device, dtype, requires_grad):
    kw = dict(device=device, dtype=dtype)
    inputs = [
        random_square_matrix_of_rank(S, S - 2, **kw),  # dim2_null
    ]
    # if you uncomment the following line, test_fn_gradgrad_linalg_det_cuda_float64 passes!
    #inputs[0] = inputs[0].cpu()
    for t in inputs:
        t.requires_grad = requires_grad
    return [SampleInput(t) for t in inputs]
```

Moreover, I tried to apply the following diff to `gradcheck`:
```diff
diff --git a/torch/autograd/gradcheck.py b/torch/autograd/gradcheck.py
index e0c48c2b60..f32631bc09 100644
--- a/torch/autograd/gradcheck.py
+++ b/torch/autograd/gradcheck.py
@@ -597,6 +597,7 @@ def gradcheck(
         inp_tensors = iter_tensors(tupled_inputs, True)
 
         for j, (a, n, inp) in enumerate(zip(analytical, numerical, inp_tensors)):
+            assert inp.is_cuda
             if a.numel() != 0 or n.numel() != 0:
                 if o.is_complex():    # C -> C, R -> C
                     if not torch.allclose(analytical_from_imag_grad_out[j], numerical_from_imag_grad_out[j], rtol, atol):
```

To summarize:
* if `sample_inputs_linalg_det` returns a `cuda` tensor, the gradcheck fails
* if it returns a `cpu` tensor, gradcheck would pass but the assert above fails, proving that sample inputs are NOT automatically converted to the correct device.

I think that this example shows that the original claim made by this issue is valid, and that there are a lot of ops for which cuda autograd is basically untested.

**Comment by antocuni** [comment]
> This issue is about test_autograd.py and not about test_ops.py. If you agree that my analysis is correct for test_autograd.py, then there is no bug and we can close this issue.

I think that the title of the issue if wrong, it should be "autograd tests", not "test_autograd". But the core fact remains: currently for many ops, we don't tests autograd on cuda.

I don't know what is the accepted workflow to handle such a case: should we just rename this issue or close this and open a new one?

> Thank you for reducing the problem to one sample_input. I am compiling your branch now and will take a look. I believe the problem is in the numerical Jacobian, it's not accurate enough (#53119 (comment)). Let's move the discussion of gradgradcheck failure for torch.linalg.det to the PR or to a separate issue.

thank you for your help on this and for the analysis!
